test/with_dummyserver/async/test_poolmanager.py

```diff
@@ -5,7 +5,7 @@
 
 class TestPoolManager:
     @conftest.test_all_backends
-    async def test_redirect(self, http_server, backend):
+    async def test_redirect(self, http_server, backend, anyio_backend):
```

Comment:
Why do we pass `anyio_backend` here if we don't use it? Asking since this will affect many tests

Reply:
Because otherwise pytest will complain. If you can find a way around this, please let me know.

Reply:
I can try asking @nicoddemus, who submitted the last hip pull request. :) My (poor) understanding is that this workaround is needed because of https://github.com/pytest-dev/pytest/commit/40072b951151ad5706c06c9d588cad087954a576. Since that code is still being worked on, maybe this issue can be fixed too.

Reply:
Hi!

That issue has been released in `5.4.1` already, but I see `hip` [pins to 4.6.6](https://github.com/python-trio/hip/blob/master/dev-requirements.txt#L7), probably to support Python 2.7 still.

Reply:
How so? Even with 5.4.1, if I omit `anyio_backend`, I get `In test_redirect: function uses no argument 'anyio_backend'`

Reply:
This argument is only needed for the the anyio pytest plugin. If there is a way to make it optional for the test function, I do not know how.

Reply:
Strange, does that message come from pytest? Can you post the full output please?

Reply:
```
==================================================================================================== test session starts ====================================================================================================
platform linux -- Python 3.6.10, pytest-5.4.1, py-1.8.0, pluggy-0.13.1
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
rootdir: /home/alex/checkout/hip, inifile: setup.cfg
plugins: trio-0.5.2, anyio-1.2.1.post3, random-order-1.0.4, flaky-3.6.1, timeout-1.3.3, cov-2.7.1
collected 0 items / 1 error                                                                                                                                                                                                 

========================================================================================================== ERRORS ===========================================================================================================
_____________________________________________________________________________ ERROR collecting test/with_dummyserver/async/test_poolmanager.py ______________________________________________________________________________
In test_redirect: function uses no argument 'anyio_backend'
================================================================================================== short test summary info ==================================================================================================
ERROR test/with_dummyserver/async/test_poolmanager.py::TestPoolManager
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
===================================================================================================== 1 error in 0.08s ======================================================================================================
```

Reply:
Ahh that's correct, that's because of `@conftest.test_all_backends`, which is defined as:

```python
test_all_backends = pytest.mark.parametrize(
    "backend, anyio_backend",
    [
        pytest.param("trio", "trio", id="trio-native"),
        pytest.param("anyio", "trio", id="anyio-trio"),
        pytest.param("anyio", "curio", id="anyio-curio"),
        pytest.param("anyio", "asyncio", id="anyio-asyncio"),
    ],
)
```

I'm not familiar with hip's codebase, but I think this means that it is expected that `backend` and `anyio_backend` to be tested together whenever `test_all_backends` is applied?

Reply:
Indeed, but @pquentin 's question was why we needed to declare the `anyio_backend` argument on the test function even though it wasn't used there.

Reply:
Hm, so the issue isn't about pytest's async test detection, but the fact that anyio added a way to control the backend which unfortunately forces every test function to add this unused arugment? Just trying to understand.

Reply:
Let me try to explain:

Here's the function definition:

```python
@conftest.test_all_backends
async def test_redirect(self, http_server, backend, anyio_backend):
    ...
```

`test_all_backends` is defined as:

```
test_all_backends = pytest.mark.parametrize(
    "backend, anyio_backend",
    [
        pytest.param("trio", "trio", id="trio-native"),
        pytest.param("anyio", "trio", id="anyio-trio"),
        pytest.param("anyio", "curio", id="anyio-curio"),
        pytest.param("anyio", "asyncio", id="anyio-asyncio"),
    ],
)
```

So effectively the function declaration is this:

```python
@pytest.mark.parametrize(
    "backend, anyio_backend",
    [
        pytest.param("trio", "trio", id="trio-native"),
        pytest.param("anyio", "trio", id="anyio-trio"),
        pytest.param("anyio", "curio", id="anyio-curio"),
        pytest.param("anyio", "asyncio", id="anyio-asyncio"),
    ],
)
async def test_redirect(self, http_server, backend, anyio_backend):
    ...
```

So, we are parametrizing with two arguments: `backend` and `anyio_backend`, so pytest rightfully complains if you are parametrizing `anyio_backend` but is not declaring the parameter. pytest can't know if you will be using it inside the function body by looking at the function signature, it is impossible.

For example, this would produce the same error:

```
@pytest.mark.parametrize('x', [1, 2, 3])
def test_foo():
    ...
```

We are parametrizing parameter `x`, but `x` is not declared in the function signature, hence the error.

Reply:
@nicoddemus Thanks, the detailed explanation makes a lot of sense, and is obvious in insight.

@agronholm Is using this introspection trick from pydantic feasible/worth it here? https://github.com/samuelcolvin/pydantic/blob/55f7d357270b37bec1483620dcb5c5360b55741b/pydantic/class_validators.py#L264-L295

Reply:
A solution would be to use `pytest.mark.usefixtures('anyio_backend')`:
```python3
def test_all_backends(func):
    func = pytest.mark.usefixtures('anyio_backend')(func)
    return pytest.mark.parametrize(
        "backend, anyio_backend",
        [
            pytest.param("trio", "trio", id="trio-native"),
            pytest.param("anyio", "trio", id="anyio-trio"),
            pytest.param("anyio", "curio", id="anyio-curio"),
            pytest.param("anyio", "asyncio", id="anyio-asyncio"),
        ]
    )(func)
```
This eliminates the need to add `anyio_backend` to every test function where we use this decorator.

Reply:
@agronholm Thanks! I tried it, but it does not seem to work. See https://github.com/pquentin/hip/commit/5a202ca04d18ae0d8ffad2ce0c1e25d4628c3572 and https://travis-ci.com/github/pquentin/hip/jobs/329567180

Reply:
Not that it's a big problem, I was just wondering if there was an easy way out

Reply:
That's weird because it works flawlessly here. I'm running latest pytest though.